### PR TITLE
treat `reallocarray` as `free`-`malloc`

### PR DESCRIPTION
**Pull request by aneksteind**
Treat `reallocarray` the same as we do `realloc`, as a `free` and then a `malloc`.

### CODE REVIEW

**Review by kkysen — CHANGES_REQUESTED** [review]
See the comment, and we should keep the snapshot up-to-date.  `cargo test` isn't tested in CI quite yet, but it will be soon.

File: analysis/runtime/src/handlers.rs [hunk 1]
```diff
@@ -63,10 +63,13 @@ pub fn realloc(mir_loc: MirLocId, old_ptr: usize, size: u64, new_ptr: usize) {
...
+    });
+    RUNTIME.send_event(Event {
+        mir_loc,
+        kind: EventKind::Alloc {
             size: (nmemb * size) as usize,
-            new_ptr,
+            ptr: new_ptr,
         },
     });
 }
```
[the remark this answers is not in the document]
Comment (aneksteind) [reply analysis/runtime/src/handlers.rs]: Let me know what you think of the changes

**Comment by kkysen** [comment]
When we hook/instrument one of these functions, where does the handler call go?  Right after the hooked function?

**Comment by kkysen** [comment]
What, for example, do we do if `malloc` returns `NULL`?

File: analysis/runtime/src/handlers.rs [hunk 2]
```diff
@@ -61,14 +62,12 @@ pub fn realloc(mir_loc: MirLocId, old_ptr: usize, size: u64, new_ptr: usize) {
...
-            old_ptr,
-            size: (nmemb * size) as usize,
-            new_ptr,
-        },
-    });
+    if let Some(total_size) = size.checked_mul(nmemb) {
+        realloc(mir_loc, old_ptr, total_size, new_ptr)
+    }
+
+    error!("Detected an overflow in reallocated size");
```
[the remark this answers is not in the document]
Comment (aneksteind) [reply analysis/runtime/src/handlers.rs]: > When we hook/instrument one of these functions, where does the handler call go? Right after the hooked function?

Good point. For hooks, it actually goes after the call -- so we'll be subject to the runtime behavior of `reallocarray` and not its hook. I'd say that the checking is best left to the eventual safe lifting, and the instrumentation shouldn't bother to check. I've pushed a commit to reflect this. https://github.com/immunant/c2rust/pull/531/commits/610452e0613b9be9c23a7b354f56661d165b9368

**Comment by aneksteind** [comment]
> What, for example, do we do if malloc returns NULL?

It would generate events for `free foo` and `malloc NULL`, which I think is fine. It's up to the instrumented program to handle that appropriately. We could change the size to zero, but it's probably helpful to track the size that was supposed to be allocated.

**Comment by kkysen** [comment]
> It would generate events for `free foo` and `malloc NULL`, which I think is fine. It's up to the instrumented program to handle that appropriately. We could change the size to zero in the event, but it's probably helpful to track the size that was supposed to be allocated.

I think we aren't quite handling things correctly still, but let's fix that in a separate PR.  I'll open an issue for it.

**Pull request merged**